### DISCUSSION TITLE
Schéma BDD actualisé : mix schémas de Nicolas et moi

### DISCUSSION THREAD

**Discussion by AdrienDSlone** [AdrienDSlone/slowTripPlanner#34] [2. Data]
J'ai actualisé le schéma de la BDD en mixant les schémas de @Scabiei et moi.

Question : une balise appartient-elle à plusieurs parcours ou bien qu'à un seul ?

## Schéma 

![20220609_BDD_actualisee](https://user-images.githubusercontent.com/50139099/172919997-301b8c71-8d00-4f74-aea7-6184ca8895bc.png)

## Code
-- MySQL Script generated by MySQL Workbench
-- Thu Jun  9 20:35:42 2022
-- Model: New Model    Version: 1.0
-- MySQL Workbench Forward Engineering

SET @OLD_UNIQUE_CHECKS=@@UNIQUE_CHECKS, UNIQUE_CHECKS=0;
SET @OLD_FOREIGN_KEY_CHECKS=@@FOREIGN_KEY_CHECKS, FOREIGN_KEY_CHECKS=0;
SET @OLD_SQL_MODE=@@SQL_MODE, SQL_MODE='ONLY_FULL_GROUP_BY,STRICT_TRANS_TABLES,NO_ZERO_IN_DATE,NO_ZERO_DATE,ERROR_FOR_DIVISION_BY_ZERO,NO_ENGINE_SUBSTITUTION';

-- -----------------------------------------------------
-- Schema mydb
-- -----------------------------------------------------

-- -----------------------------------------------------
-- Schema mydb
-- -----------------------------------------------------
CREATE SCHEMA IF NOT EXISTS `mydb` DEFAULT CHARACTER SET utf8mb3 ;
USE `mydb` ;

-- -----------------------------------------------------
-- Table `mydb`.`moyen_locomotion`
-- -----------------------------------------------------
CREATE TABLE IF NOT EXISTS `mydb`.`moyen_locomotion` (
  `id` INT NOT NULL,
  `intitule` VARCHAR(45) NULL DEFAULT NULL,
  PRIMARY KEY (`id`))
ENGINE = InnoDB
DEFAULT CHARACTER SET = utf8mb3;

-- -----------------------------------------------------
-- Table `mydb`.`difficulte`
-- -----------------------------------------------------
CREATE TABLE IF NOT EXISTS `mydb`.`difficulte` (
  `id` INT NOT NULL,
  `intitule` VARCHAR(45) NOT NULL,
  PRIMARY KEY (`id`))
ENGINE = InnoDB
DEFAULT CHARACTER SET = utf8mb3;

-- -----------------------------------------------------
-- Table `mydb`.`preference`
-- -----------------------------------------------------
CREATE TABLE IF NOT EXISTS `mydb`.`preference` (
  `id` INT NOT NULL,
  `intitulé` VARCHAR(45) NOT NULL,
  `locomotion_id` INT NOT NULL,
  `difficulte_id` INT NOT NULL,
  `km_moyen_par_etape` FLOAT NULL DEFAULT NULL,
  `km_max_jour` FLOAT NULL DEFAULT NULL,
  `vitesse_desiree` FLOAT NULL DEFAULT NULL,
  `frequence_repos` TIME NULL DEFAULT NULL,
  `duree_repos` TIME NULL DEFAULT NULL,
  `frequence_pause` TIME NULL DEFAULT NULL,
  `duree_pause` TIME NULL DEFAULT NULL,
  `poids_max` FLOAT NULL DEFAULT NULL,
  PRIMARY KEY (`id`, `locomotion_id`, `difficulte_id`),
  INDEX `fk_preference_locomotion1_idx` (`locomotion_id` ASC) VISIBLE,
  INDEX `fk_preference_difficulte1_idx` (`difficulte_id` ASC) VISIBLE,
  CONSTRAINT `fk_preference_difficulte1`
    FOREIGN KEY (`difficulte_id`)
    REFERENCES `mydb`.`difficulte` (`id`),
  CONSTRAINT `fk_preference_locomotion1`
    FOREIGN KEY (`locomotion_id`)
    REFERENCES `mydb`.`moyen_locomotion` (`id`))
ENGINE = InnoDB
DEFAULT CHARACTER SET = utf8mb3;

-- -----------------------------------------------------
-- Table `mydb`.`parcours`
-- -----------------------------------------------------
CREATE TABLE IF NOT EXISTS `mydb`.`parcours` (
  `id` INT NOT NULL,
  `nom` VARCHAR(100) NULL DEFAULT NULL,
  `date_depart` DATE NOT NULL,
  `date_arrivee` DATE NULL DEFAULT NULL,
  `point_depart_lattitude` DOUBLE NULL DEFAULT NULL,
  `point_depart_longitude` DOUBLE NULL DEFAULT NULL,
  `point_depart_intitule` VARCHAR(45) NULL DEFAULT NULL,
  `point_arrivee_lattitude` DOUBLE NULL DEFAULT NULL,
  `point_arrivee_longitude` DOUBLE NULL DEFAULT NULL,
  `point_arrivee_intitule` VARCHAR(45) NULL DEFAULT NULL,
  `description` VARCHAR(255) NULL DEFAULT NULL,
  `preference_id` INT NOT NULL,
  PRIMARY KEY (`id`, `preference_id`),
  INDEX `fk_parcours_preference1_idx` (`preference_id` ASC) VISIBLE,
  CONSTRAINT `fk_parcours_preference1`
    FOREIGN KEY (`preference_id`)
    REFERENCES `mydb`.`preference` (`id`))
ENGINE = InnoDB
DEFAULT CHARACTER SET = utf8mb3;

-- -----------------------------------------------------
-- Table `mydb`.`balise`
-- -----------------------------------------------------
CREATE TABLE IF NOT EXISTS `mydb`.`balise` (
  `id` INT NOT NULL,
  `nom` VARCHAR(45) NULL,
  `date` DATE NOT NULL,
  `latitude` DOUBLE NOT NULL,
  `longitude` DOUBLE NOT NULL,
  `distance` INT NOT NULL,
  `positionDansParcours` INT NOT NULL,
  `isEtape` TINYINT NOT NULL,
  `intitule` VARCHAR(45) NULL DEFAULT NULL,
  `heureArrivee` DATETIME NULL DEFAULT NULL,
  `heureDepart` DATETIME NULL DEFAULT NULL,
  `description` VARCHAR(255) NULL,
  `est_verrouillee` TINYINT NULL,
  `parcours_id` INT NOT NULL,
  `locomotion_id` INT NOT NULL,
  PRIMARY KEY (`id`, `parcours_id`, `locomotion_id`),
  INDEX `fk_balise_parcours1_idx` (`parcours_id` ASC) VISIBLE,
  INDEX `fk_balise_locomotion1_idx` (`locomotion_id` ASC) VISIBLE,
  CONSTRAINT `fk_balise_locomotion1`
    FOREIGN KEY (`locomotion_id`)
    REFERENCES `mydb`.`moyen_locomotion` (`id`),
  CONSTRAINT `fk_balise_parcours1`
    FOREIGN KEY (`parcours_id`)
    REFERENCES `mydb`.`parcours` (`id`))
ENGINE = InnoDB
DEFAULT CHARACTER SET = utf8mb3;

-- -----------------------------------------------------
-- Table `mydb`.`equipement`
-- -----------------------------------------------------
CREATE TABLE IF NOT EXISTS `mydb`.`equipement` (
  `id` INT NOT NULL,
  `intitule` VARCHAR(45) NOT NULL,
  `poids` FLOAT NULL DEFAULT NULL,
  `description` VARCHAR(100) NULL DEFAULT NULL,
  PRIMARY KEY (`id`))
ENGINE = InnoDB
DEFAULT CHARACTER SET = utf8mb3;

-- -----------------------------------------------------
-- Table `mydb`.`equipement_has_parcours`
-- -----------------------------------------------------
CREATE TABLE IF NOT EXISTS `mydb`.`equipement_has_parcours` (
  `equipement_id` INT NOT NULL,
  `parcours_id` INT NOT NULL,
  PRIMARY KEY (`equipement_id`, `parcours_id`),
  INDEX `fk_equipement_has_parcours_parcours1_idx` (`parcours_id` ASC) VISIBLE,
  INDEX `fk_equipement_has_parcours_equipement1_idx` (`equipement_id` ASC) VISIBLE,
  CONSTRAINT `fk_equipement_has_parcours_equipement1`
    FOREIGN KEY (`equipement_id`)
    REFERENCES `mydb`.`equipement` (`id`),
  CONSTRAINT `fk_equipement_has_parcours_parcours1`
    FOREIGN KEY (`parcours_id`)
    REFERENCES `mydb`.`parcours` (`id`))
ENGINE = InnoDB
DEFAULT CHARACTER SET = utf8mb3;

-- -----------------------------------------------------
-- Table `mydb`.`utilisateur`
-- -----------------------------------------------------
CREATE TABLE IF NOT EXISTS `mydb`.`utilisateur` (
  `id` INT NOT NULL,
  `nom` VARCHAR(45) NOT NULL,
  `prenom` VARCHAR(45) NOT NULL,
  `email` VARCHAR(45) NOT NULL,
  `motdepasse` VARCHAR(45) NOT NULL,
  `urlPhoto` VARCHAR(45) NULL DEFAULT NULL,
  PRIMARY KEY (`id`))
ENGINE = InnoDB
DEFAULT CHARACTER SET = utf8mb3;

-- -----------------------------------------------------
-- Table `mydb`.`equipement_has_utilisateur`
-- -----------------------------------------------------
CREATE TABLE IF NOT EXISTS `mydb`.`equipement_has_utilisateur` (
  `equipement_id` INT NOT NULL,
  `utilisateur_id` INT NOT NULL,
  PRIMARY KEY (`equipement_id`, `utilisateur_id`),
  INDEX `fk_equipement_has_utilisateur_utilisateur1_idx` (`utilisateur_id` ASC) VISIBLE,
  INDEX `fk_equipement_has_utilisateur_equipement1_idx` (`equipement_id` ASC) VISIBLE,
  CONSTRAINT `fk_equipement_has_utilisateur_equipement1`
    FOREIGN KEY (`equipement_id`)
    REFERENCES `mydb`.`equipement` (`id`),
  CONSTRAINT `fk_equipement_has_utilisateur_utilisateur1`
    FOREIGN KEY (`utilisateur_id`)
    REFERENCES `mydb`.`utilisateur` (`id`))
ENGINE = InnoDB
DEFAULT CHARACTER SET = utf8mb3;

-- -----------------------------------------------------
-- Table `mydb`.`parcours_has_utilisateur`
-- -----------------------------------------------------
CREATE TABLE IF NOT EXISTS `mydb`.`parcours_has_utilisateur` (
  `parcours_id` INT NOT NULL,
  `utilisateur_id` INT NOT NULL,
  PRIMARY KEY (`parcours_id`, `utilisateur_id`),
  INDEX `fk_parcours_has_utilisateur_utilisateur1_idx` (`utilisateur_id` ASC) VISIBLE,
  INDEX `fk_parcours_has_utilisateur_parcours_idx` (`parcours_id` ASC) VISIBLE,
  CONSTRAINT `fk_parcours_has_utilisateur_parcours`
    FOREIGN KEY (`parcours_id`)
    REFERENCES `mydb`.`parcours` (`id`),
  CONSTRAINT `fk_parcours_has_utilisateur_utilisateur1`
    FOREIGN KEY (`utilisateur_id`)
    REFERENCES `mydb`.`utilisateur` (`id`))
ENGINE = InnoDB
DEFAULT CHARACTER SET = utf8mb3;

SET SQL_MODE=@OLD_SQL_MODE;
SET FOREIGN_KEY_CHECKS=@OLD_FOREIGN_KEY_CHECKS;
SET UNIQUE_CHECKS=@OLD_UNIQUE_CHECKS;

**Comment by jdenizart** [comment]
EQUIPEMENT:
Je pense que l'équipement peut etre interessant pour un voyageur et sera un élément à rajouter sur un site mais ce n'est pas la priorite, car si on entre dedans, on peut developper des bases enormes.  Avec de nombreuses sous-categories. et pour chaque article, un type, le modele, la marque, une description, un poids, une quantité, éventuellement un prix...
voir la liste sur le liste ou sur un de mes fichiers excels 
https://www.cyclo-randonnee.fr/lire/rub-liste-materiel-49.html

**Reply by AdrienDSlone** [reply]
On peut présenter la base de données ideales pour montrer que le projet a été conçu pour avoir un avenir. Sinon on ne saura pas quoi repondre à la question " vous envisagez des equipments mais vous avez pas prevu de table... C'est normal ?"

**Comment by jdenizart** [comment]
PREFERENCE:
Pour moi, ce n'est qu'une extension du parcours, donc un parcours a une seule pref, et une pref est lié a un seul parcours.

**Reply by AdrienDSlone** [reply]
On pourrait vouloir appliquer nos préférences sauvegardées à chaque création de parcours. 

Si les preferences sont propres au parcours, il faudra sauvegarder ces infos dans l'entitée parcours, ca signifierait qu'on y stockerait prariquement tout. On aurait Parcours, balise et utilisateur, ça me semble très peu modulable.

**Comment by jdenizart** [comment]
BALISE:
1/- Il serait peut etre interessant de diviser cette table en deux car elle peut contenir beaucoup de données.
2/ - du fait du nombre de parametres et qui sont sont en plus très ouvert, qui definissent une balise, pour moi il est quasi impossible  d'avoir deux meme balises. donc j'auai tendance à dire que c'est une relation 1,N (en fait N,1) et pas N,N avec le parcours

**Reply by AdrienDSlone** [reply]
Le schema contient bien une relation 1-n, plusieurs balisent appartiennent à un parcours

**Comment by jdenizart** [comment]
PARCOURS
je pense qu'il y a de nombreux champs qui sont simplement un calcul. Doivent ils etre persistant? je le pense quand meme. Pour les coordonnées du depart et arrivée, je pense que ce n'est pas interessant, mais juste recuperer le nom du point.

**Reply by AdrienDSlone** [reply]
D'accord avec ça, meme pas besoin de nom, la balise le fait pour nous

**Comment by jdenizart** [comment]
DIFFICULTE:
 facile, moyen, difficile
ou débutant, normal, expert ou confirmé
ou je me ballade tout doux (tranquillou avec des visites), on roule , je me suis lancé un defi chrono/je perds pas de temps

**Reply by AdrienDSlone** [reply]
Donc un id et un libellé, c'est le libellé qui précisera ce qu'on choisit

**Reply by AdrienDSlone** [reply]
Pour l'echelle, je te rejoins autant avoir cinq ou six niveaux de difficulté c'est plus personnalisable

**Comment by jdenizart** [comment]
MOYEN de LOCOMOTION:
oui il faut les relier à la balise et aux preferences

**Comment by jdenizart** [comment]
![SlowTripPlanner - BDD - Schema - Maximale Ultime](https://user-images.githubusercontent.com/104148814/176919177-5c39486a-bac6-4918-a82c-4415c4b8b13c.png)
La version la plus complete!!!!
